### ISSUE TITLE
Formatter should not join lines that were splitted purposedely

### ISSUE THREAD

**Issue by rkj** [dart-lang/dart_style#229]
Example constructor declaration:

    SomeLongNameForAComponent(DirectiveInjector injector,
        RenderSyncZoneAccess zoneAccess,
        this._formatters,
        this._element,
        this._scope);

gets formatted into:

    SomeLongNameForAComponent(DirectiveInjector injector,
        RenderSyncZoneAccess zoneAccess, this._formatters, this._element,
        this._scope);

**Comment by munificent** [comment]
Yes, this is similar to #8 and #141.

I'm starting to work on this but, strangely enough, it requires fairly deep structural changes in the formatter to handle. Its current rule engine basically only handles binary constraints and it needs a more rule-based approach to handle things like argument and parameter lists where there are a finite set of "patterns" for formatting it.

**Comment by rkj** [comment]
Thanks a lot for working on it, I have no doubt that it's not easy.

Another example to consider:

    class _MultiSelectionModelImpl<T>
        extends ChangeNotifier
        with SelectionChangeNotifier<T>
        implements SelectionModel<T> {

gets formatted into:

    class _MultiSelectionModelImpl<T> extends ChangeNotifier
        with SelectionChangeNotifier<T> implements MultiSelectionModel<T> {

that is not too bad, but I still like the original better.

**Comment by rkj** [comment]
Here's another, taken from https://angulardart.org/tutorial/05-ch03-component.html
```
@Component(
    selector: 'rating',
    templateUrl: 'rating.html',
    cssUrl: 'rating.css')
class RatingComponent {}   
```
gets squeezed into:
```
@Component(selector: 'rating', templateUrl: 'rating.html', cssUrl: 'rating.css')
class RatingComponent {}
```

**Comment by munificent** [comment]
Can you file a separate bug for [this](https://github.com/dart-lang/dart_style/issues/229#issuecomment-89440152), please? I'm on the fence about how the formatter should do this, but I think your preference might be acceptable. I just have to see how it affects other examples.

Your [other comment](https://github.com/dart-lang/dart_style/issues/229#issuecomment-89449378) is as-designed. If the argument list does fit on one line, it will put it on one.

**Comment by munificent** [comment]
Fixed by 346ad0736b5aff21ce5751c818bd4ccfc02d5885!
